### PR TITLE
JavaScript: Add getter summaries and receiver flow.

### PR DESCRIPTION
**Pull request by xiemaisi**
This PR adds two new extensions to the data-flow library:

  1. We now track data and taint through getter functions that return a property of one of their arguments. (We already did the same for setters.)

  2. We now track data and taint from method receivers into the `this` of their target.

Performance worked out [pretty well](https://git.semmle.com/max/dist-compare-reports/blob/master/js/getter-summaries/report.md) (internal link) considering that both of these extensions are non-trivial. While `aura` and `ChromeFuzzer` are slightly above the magical 5% threshold, I think it's unlikely we'll get everything below 5% for a deep change like this, and I think the new results are worth a minor slowdown: the two new results on amphtml, in particular, look pleasingly non-trivial and plausible (though they are in the build system, so not exploitable).

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/dataflow/Configuration.qll [hunk 1]
```diff
@@ -795,7 +795,7 @@ private predicate summarizedHigherOrderCall(
 ) {
   exists(
     Function f, DataFlow::InvokeNode outer, DataFlow::InvokeNode inner, int j,
-    DataFlow::Node innerArg, DataFlow::ParameterNode cbParm, PathSummary oldSummary
+    DataFlow::Node innerArg, DataFlow::SourceNode cbParm, PathSummary oldSummary
```
Comment (asger-semmle) [inline javascript/ql/src/semmle/javascript/dataflow/Configuration.qll]: Could you explain this change? It seems unlikely that `this` would ever be used as a callback.

File: javascript/ql/src/semmle/javascript/dataflow/internal/FlowSteps.qll [hunk 2]
```diff
@@ -125,20 +125,27 @@ private module CachedSteps {
...
-      arg = invk.getArgument(i)
+    (
+      exists(int i, Parameter p |
+        f.getParameter(i) = p and
+        not p.isRestParameter() and
+        arg = invk.getArgument(i) and
+        parm = DataFlow::parameterNode(p)
+      )
+      or
+      arg = invk.(DataFlow::MethodCallNode).getReceiver() and
```
Comment (asger-semmle) [inline javascript/ql/src/semmle/javascript/dataflow/internal/FlowSteps.qll]: If we use `invk.(DataFlow::CallNode).getReceiver()` here we'll also handle receivers passed in through reflective calls.

File: javascript/ql/src/semmle/javascript/dataflow/Configuration.qll (hunk 1 above)
Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/dataflow/Configuration.qll]: True, it's unlikely. I went through and changed all places where the final argument of `argumentPassing` was implicitly cast to a `ParameterNode` and generalised it to `SourceNode`. If you like I can revert this particular change.

File: javascript/ql/src/semmle/javascript/dataflow/internal/FlowSteps.qll (hunk 2 above)
Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/dataflow/internal/FlowSteps.qll]: Ah, good point. This was simply a typo, I think. I'll do a quick sanity check to verify that this doesn't upset the optimiser.

File: javascript/ql/src/semmle/javascript/dataflow/Configuration.qll (hunk 1 above)
Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/dataflow/Configuration.qll]: I see. Probably doesn't matter, so no need to revert if it's working as it is.

File: javascript/ql/src/semmle/javascript/dataflow/internal/FlowSteps.qll (hunk 2 above)
Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/dataflow/internal/FlowSteps.qll]: Sanity check showed no performance regression.

**Pull request merged**